### PR TITLE
Add jstack (8,11,12) and update jps (8,11)

### PR DESCRIPTION
**Pull request by SueChaplain**
- New tools topic for jstack
- New section for jstack in the release topic
- Updated to Tools "Switching to OpenJ9" to describe differences to
  HotSpot jstack

- Add jps to release topic for 8 & 11 (12 was in 0.13.0)
- Remove icons from jps content that specify Java 12 only.

Closes: #225

Signed-off-by: Sue Chaplain <sue_chaplain@uk.ibm.com>

### CODE REVIEW

**Comment by SueChaplain** [comment]
Jenkins doc stage

**Comment by pshipton** [comment]
Note there is a merge conflict.

**Comment by SueChaplain** [comment]
Jenkins doc stage

File: docs/tool_jstack.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+    jstack <options>* <pid>*
+
+Where `<pid>*` is a list of process IDs. If none are supplied, the process IDs are read from `stdin`, which allows a user to query all processes via `jps -q | jstack`. IDs of inactive processes are silently ignored. If this option is used on its own, the output contains Java stacks and thread information of the specified processes (equivalent to the information provided in `java.lang.managmenet.ThreadInfo`).
+
+The values for `<options>*` are as follows:
+
+- `-p`: Prints the system and agent properties of the process
+- `-J`: supplies arguments to the Java VM that is running the `jstack` command. You can use multiple `-J` options, for example: `jstack -J-Xmx10m -J-Dcom.ibm.tools.attach.enable=yes`
+- `-l`: prints the application package name
```
Comment (pshipton) [inline docs/tool_jstack.md]: According to the jstack help output, these are the only options supported
```
				+ "    -p: print the target's system and agent properties%n"
				+ "    -l: Long format. Print the thread's ownable synchronizers%n"
				+ "    -J: supply arguments to the Java VM running jps%n"
```

**Comment by SueChaplain** [comment]
@pdbain-ibm @pshipton 

Please can this be reviewed your time today so that any changes I need to make can happen tomorrow, after which I am on vacation for a week.

I'm unsure about the list of options, because the issue says *The usual suspects: -J, -h, like jps.*

I've also added some references to Linux in there where I think options/syntax is not possible on Windows platforms. Might need some tweaks.

I've staged the doc but the latest updates are still building, so you'll need to wait for ghpages to refresh before checking: https://eclipse.github.io/openj9-docs-staging/246/tool_jstack/

I changed the title of the topic to "Java stack tool", because it doesn't look good in the table of contents, so when you see that change you'll know its updated.

**Comment by pdbain-ibm** [comment]
@SueChaplain Thanks for working on this!

> I think options/syntax is not possible on Windows platforms

You could say "user __running a Bourne or equivalent shell__ to..."

Please drop these options:
```
-m: prints the application arguments
-v: prints the Java VM arguments, including those that are produced automatically
```
and add the same text for `-h` as `jps`.

`If this option is used on its own,`: I am not sure what you mean by this.  I don't believe it is necessary.

> I changed the title of the topic to "Java stack tool", because it doesn't look good in the table of contents, so when you see that change you'll know its updated.

Could we include the actual tool name table of contents, e.g. `Java stack tool (jstack)` or `jstack: Java stack tool`, or other variant of these (same for `jps`). This makes it clear we have these tools and helps search engines find them.

For reference:
https://docs.oracle.com/javase/7/docs/technotes/tools/share/jstack.html is titled `jstack - Stack Trace` and https://docs.oracle.com/javase/7/docs/technotes/tools/share/jps.html is titled `jps - Java Virtual Machine Process Status Tool`.  For reference also: https://docs.oracle.com/javase/7/docs/technotes/tools/

**Comment by DanHeidinga** [comment]
@SueChaplain Oracle makes it very clear that the j* tools are 
> are unsupported and experimental, and should be used with that in mind. They may not be available in future JDK versions.

We should include similar restrictions in openj9 documentation for the these tools

File: docs/tool_jstack.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+- `-q`: prints only the virtual machine identifiers
+- `-m`: prints the application arguments
+- `-v`: prints the Java VM arguments, including those that are produced automatically
+
+
+<i class="fa fa-exclamation-triangle" aria-hidden="true"></i> **Restrictions:**
```
Comment (pshipton) [inline docs/tool_jstack.md]: > We should include similar restrictions in openj9 documentation for the these tools

@DanHeidinga is this what you are looking for, or does it need to be clarified or stated in other places as well?

Comment (DanHeidinga) [reply docs/tool_jstack.md]: Yup, that's exactly what I was looking for.  Missed it on first read

**Comment by SueChaplain** [comment]
@pdbain-ibm 

Some confusion over options still:

>add the same text for `-h` as jps

There is no `-h` option for `jps`. I assume you mean that you have added a `-h` to print help information, to make this compatible with `jstack`.  I added this.

>Please drop these options: (-m -v)

Done.

So, with these corrected we still have some discrepancies:

| HotSpot option | OpenJ9 option |
|-------------------|-----------------|
| -J | -J |
| -l | -l |
| -h | -h |
| -help | - |
| -F| not supported |
| -m| native info not supported|
|- | -p |
| -| -q |

You  gave us the `-p` option in your issue, but not `-q`, which got copied from `jps` because of the *all the usual suspects* comment. I assume we remove `-q` here (which I have done in my latest update)? Please confirm. Please also confirm that `-p` is correct.

Forgive me for not looking directly at the Oracle documentation when writing this up. We deliberately avoid doing so because of copyright infringement. It is therefore important that we have all the information directly from you.

>If this option is used on its own,: I am not sure what you mean by this. I don't believe it is necessary.

I meant if it is used as `jstack <pid*>` without adding any `<options>`. I removed the words anyway as the sentence is correct without them.

>Could we include the actual tool name table of contents

I played around with this a bit yesterday. I do agree with you. On their own they look odd. Although we try to avoid long TOC titles, we have longer ones, so I've added in *Full name (cmdname)* for `jstack` and `jps`. Hope that works for you.

**Comment by SueChaplain** [comment]
Jenkins doc stage

File: docs/tool_jstack.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+# The jstack tool
+
+Use the `jstack` tool to obtain Java stack traces and thread information for processes. The tool is similar to the HotSpot tool of the same name; the OpenJ9 version of `jstack` is an independent implementation, added for compatibility.
+
+The command syntax is as follows:
+
+    jstack <options>* <pid>*
+
+Where `<pid>*` is a list of process IDs. If none are supplied, the process IDs are read from `stdin`, which allows a user to query all processes via `jps -q | jstack`. IDs of inactive processes are silently ignored. If this option is used on its own, the output contains Java stacks and thread information of the specified processes (equivalent to the information provided in `java.lang.managmenet.ThreadInfo`).
```
[the remark this answers is not in the document]
Comment (SueChaplain) [reply docs/tool_jstack.md]: Fixed

**Comment by SueChaplain** [comment]
Jenkins doc stage

**Comment by SueChaplain** [comment]
Jenkins doc stage

**Comment by pdbain-ibm** [comment]
@SueChaplain 
Thanks for the updates. Reviewed and approved.

**Comment by SueChaplain** [comment]
@pshipton - if you are also ok with the updates, please go ahead and merge.

**Comment by SueChaplain** [comment]
@pshipton I'm out until after MS1, so if there are any other changes needed these will have to be done on top of my merge, or in a separate PR. Thanks!

**Pull request merged**